File: src/game_learning_runtime/season.py
"""Declarative season selection and portable, byte-bound role context."""

from __future__ import annotations

import hashlib
import json
import os
import re
import sys
import tempfile
from collections.abc import Mapping
from dataclasses import dataclass
from pathlib import Path
from types import MappingProxyType
from typing import TYPE_CHECKING, Any

if sys.version_info >= (3, 11):
    import tomllib
else:
    import tomli as tomllib

if TYPE_CHECKING:
    from game_learning_runtime.project import GLRProject

CONTEXT_SCHEMA = "glr.season-context.v1"
ENVIRONMENT_KEYS = (
    "GLR_SEASON_ID",
    "GLR_RULESET_ID",
    "GLR_SEASON_CONFIG_SHA256",
    "GLR_SEASON_CONTEXT_SHA256",
    "GLR_SEASON_CONTEXT",
)
_LIMIT = 1024 * 1024
_IDENTIFIER = re.compile(r"^[a-z][a-z0-9_-]{0,63}$")


def _identifier(value: object) -> str:
    if not isinstance(value, str) or not _IDENTIFIER.fullmatch(value):
        raise ValueError("season identifiers must be portable lowercase identifiers (1..64)")
    return value


def _strict(value: object, required: set[str], optional: set[str] | None = None) -> dict[str, Any]:
    if (
        not isinstance(value, dict)
        or set(value) - required - (optional or set())
        or required - set(value)
    ):
        raise ValueError("season configuration has missing or unknown fields")
    return value


def season_path(root: Path, relative: object) -> Path:
    """Validate every path component, including links and Windows reparse points."""
    if (
        not isinstance(relative, str)
        or len(relative.encode("utf-8")) > 512
        or any(c in relative for c in "\\:\x00")
    ):
        raise ValueError("season config must be a portable root-relative path")
    if any(not part or part in {".", ".."} for part in relative.split("/")):
        raise ValueError("season config must be a portable root-relative path")
    candidate = root.resolve()
    for part in relative.split("/"):
        if any(ord(c) < 32 or 127 <= ord(c) <= 159 for c in part):
            raise ValueError("season config path contains control characters")
        candidate /= part
        if candidate.is_symlink() or (
            candidate.exists() and getattr(candidate.lstat(), "st_file_attributes", 0) & 0x400
        ):
            raise ValueError("season config cannot contain links or reparse points")
    if not candidate.resolve().is_relative_to(root.resolve()):
        raise ValueError("season config escapes project root")
    return candidate


def _canonical(value: Mapping[str, Any]) -> str:
    return json.dumps(value, sort_keys=True, ensure_ascii=False, separators=(",", ":"))


def _unique_object(pairs: list[tuple[str, Any]]) -> dict[str, Any]:
    result: dict[str, Any] = {}
    for key, value in pairs:
        if key in result:
            raise ValueError("duplicate key in frozen season JSON")
        result[key] = value
    return result


@dataclass(frozen=True, slots=True)
class SeasonInput:
    path: str
    sha256: str
    size_bytes: int

    def to_mapping(self) -> dict[str, Any]:
        return {"path": self.path, "sha256": self.sha256, "size_bytes": self.size_bytes}


def _read(root: Path, relative: str) -> tuple[SeasonInput, bytes]:
    path = season_path(root, relative)
    if not path.is_file():
        raise ValueError("season config must be an existing regular file")
    with path.open("rb") as stream:
        raw = stream.read(_LIMIT + 1)
    if len(raw) > _LIMIT:
        raise ValueError("season input exceeds the 1 MiB limit")
    return SeasonInput(relative, hashlib.sha256(raw).hexdigest(), len(raw)), raw


def _toml(root: Path, relative: str) -> tuple[SeasonInput, dict[str, Any]]:
    if not relative.endswith(".toml"):
        raise ValueError("season declarations and catalogs must be TOML")
    reference, raw = _read(root, relative)
    return reference, tomllib.loads(raw.decode("utf-8"))


@dataclass(frozen=True, slots=True)
class SeasonContext:
    season_id: str
    ruleset_id: str
    environment_id: str
    protocol_version: str
    status: str
    project: SeasonInput
    catalog: SeasonInput
    declaration: SeasonInput
    extensions: Mapping[str, SeasonInput]

    def to_mapping(self) -> dict[str, Any]:
        value = {
            "schema_version": CONTEXT_SCHEMA,
            "season_id": self.season_id,
            "ruleset_id": self.ruleset_id,
            "environment_id": self.environment_id,
            "protocol_version": self.protocol_version,
            "status": self.status,
            "project": self.project.to_mapping(),
            "catalog": self.catalog.to_mapping(),
            "declaration": self.declaration.to_mapping(),
            "extensions": {key: item.to_mapping() for key, item in self.extensions.items()},
        }
        value["context_sha256"] = hashlib.sha256(_canonical(value).encode("utf-8")).hexdigest()
        return value

    def to_json(self) -> str:
        result = _canonical(self.to_mapping())
        if len(result.encode("utf-8")) > 24 * 1024:
            raise ValueError("season context exceeds the 24 KiB role-environment limit")
        return result

    def require_ready(self) -> None:
        if self.status != "ready":
            raise ValueError(
                "selected season is pending; train/goal/play require ready configuration"
            )

    def verify(self, root: Path) -> None:
        for reference in (self.project, self.catalog, self.declaration, *self.extensions.values()):
            if _read(root, reference.path)[0] != reference:
                raise ValueError("frozen season input changed before role execution")

    def environment(self, root: Path) -> dict[str, str]:
        self.verify(root)
        return dict(
            zip(
                ENVIRONMENT_KEYS,
                (
                    self.season_id,
                    self.ruleset_id,
                    self.declaration.sha256,
                    self.to_mapping()["context_sha256"],
                    self.to_json(),
                ),
                strict=True,
            )
        )


def _catalog(project: GLRProject) -> tuple[SeasonInput, list[dict[str, Any]]]:
    if project.seasons is None:
        raise ValueError("project has no [seasons] config reference")
    reference, value = _toml(project.root, project.seasons)
    _strict(value, {"schema_version", "entries"})
    if value["schema_version"] != "glr.seasons.v1" or not isinstance(value["entries"], list):
        raise ValueError("unsupported season catalog schema")
    entries = value["entries"]
    if len(entries) > 256:
        raise ValueError("season catalog exceeds 256 entries")
    pairs: set[tuple[str, str]] = set()
    for entry in entries:
        _strict(entry, {"season_id", "ruleset_id", "config"})
        pair = (_identifier(entry["season_id"]), _identifier(entry["ruleset_id"]))
        if pair in pairs:
            raise ValueError("duplicate season/ruleset selection")
        pairs.add(pair)
        season_path(project.root, entry["config"])
    return reference, entries


def select_season(
    project: GLRProject, season: str | None, ruleset: str | None
) -> SeasonContext | None:
    if (season is None) != (ruleset is None):
        raise ValueError("--season and --ruleset must be supplied together")
    if season is None:
        return None
    _identifier(season)
    _identifier(ruleset)
    catalog, entries = _catalog(project)
    entry = next(
        (item for item in entries if (item["season_id"], item["ruleset_id"]) == (season, ruleset)),
        None,
    )
    if entry is None:
        raise ValueError("unknown season/ruleset selection")
    declaration, value = _toml(project.root, entry["config"])
    _strict(
        value,
        {
            "schema_version",
            "season_id",
            "ruleset_id",
            "environment_id",
            "protocol_version",
            "status",
        },
        {"extensions"},
    )
    if (
        value["schema_version"],
        value["season_id"],
        value["ruleset_id"],
        value["environment_id"],
        value["protocol_version"],
    ) != (
        "glr.season.v1",
        season,
        ruleset,
        project.environment_id,
        project.protocol_version,
    ):
        raise ValueError("season declaration identity does not match project and selection")
    if value["status"] not in {"pending", "ready"}:
        raise ValueError("season status must be pending or ready")
    extensions = value.get("extensions", {})
    if not isinstance(extensions, dict) or len(extensions) > 32:
        raise ValueError("season extensions must be a table with at most 32 entries")
    references = {}
    for namespace, config in extensions.items():
        _identifier(namespace)
        _strict(config, {"config"})
        references[namespace] = _read(project.root, config["config"])[0]
    assert project.manifest_path is not None
    manifest = _read(project.root, project.manifest_path.relative_to(project.root).as_posix())[0]
    if manifest.sha256 != project.manifest_sha256:
        raise ValueError("project manifest changed while selecting season")
    assert ruleset is not None
    return SeasonContext(
        season,
        ruleset,
        project.environment_id,
        project.protocol_version,
        value["status"],
        manifest,
        catalog,
        declaration,
        MappingProxyType(references),
    )


def list_seasons(project: GLRProject) -> list[dict[str, Any]]:
    if project.seasons is None:
        return []
    return [
        select_season(project, item["season_id"], item["ruleset_id"]).to_mapping()  # type: ignore[union-attr]
        for item in _catalog(project)[1]
    ]


def require_selection(project: GLRProject, *, ready: bool) -> None:
    if project.seasons is not None and project.season_context is None:
        raise ValueError("project requires explicit --season and --ruleset before role execution")
    if project.season_context is not None:
        project.season_context.verify(project.root)
        if ready:
            project.season_context.require_ready()


def load_season_context(
    project: GLRProject, environment: Mapping[str, str] | None = None
) -> SeasonContext | None:
    """Revalidate the CLI's frozen bytes, not a project-local active-profile fallback."""
    source = os.environ if environment is None else environment
    present = [key for key in ENVIRONMENT_KEYS if key in source]
    if not present:
        return None
    if len(present) != len(ENVIRONMENT_KEYS) or len(source["GLR_SEASON_CONTEXT"]) > 64 * 1024:
        raise ValueError("incomplete or oversized frozen season environment")
    context = select_season(project, source["GLR_SEASON_ID"], source["GLR_RULESET_ID"])
    assert context is not None
    expected = context.environment(project.root)
    if _canonical(
        json.loads(source["GLR_SEASON_CONTEXT"], object_pairs_hook=_unique_object)
    ) != context.to_json() or any(
        source[key] != expected[key] for key in ENVIRONMENT_KEYS if key != "GLR_SEASON_CONTEXT"
    ):
        raise ValueError("CLI frozen season context does not match current inputs")
    return context


def initialize_season(project: GLRProject, season: str, ruleset: str) -> SeasonContext:
    """Create only pending TOML and register it; never overwrite or execute a hook."""
    season, ruleset = _identifier(season), _identifier(ruleset)
    if project.seasons is None:
        raise ValueError('first declare [seasons] config = "config/seasons.toml" in the project')
    catalog_path = season_path(project.root, project.seasons)
    if not project.seasons.endswith(".toml"):
        raise ValueError("season catalog must be TOML")
    original = _read(project.root, project.seasons)[1] if catalog_path.exists() else None
    entries = _catalog(project)[1] if original is not None else []
    if any((item["season_id"], item["ruleset_id"]) == (season, ruleset) for item in entries):
        raise FileExistsError("season/ruleset already registered")
    if len(entries) >= 256:
        raise ValueError("season catalog exceeds 256 entries")
    relative = f"config/seasons/{season}/{ruleset}.toml"
    if relative == project.seasons:
        raise ValueError("catalog and declaration paths must differ")
    declaration = season_path(project.root, relative)
    if declaration == catalog_path:
        raise ValueError("catalog and declaration paths must differ")
    declaration.parent.mkdir(parents=True, exist_ok=True)
    catalog_path.parent.mkdir(parents=True, exist_ok=True)
    lock = catalog_path.with_name(catalog_path.name + ".lock")
    lock_stream = lock.open("xb")
    try:
        season_path(project.root, project.seasons)
        current = _read(project.root, project.seasons)[1] if catalog_path.exists() else None
        if current != original:
            raise ValueError("season catalog changed during initialization")
        # Serialize only the validated catalog schema, preserving existing entry order.
        catalog_text = 'schema_version = "glr.seasons.v1"\n'
        for item in [*entries, {"season_id": season, "ruleset_id": ruleset, "config": relative}]:
            catalog_text += "\n[[entries]]\n" + "".join(
                f"{key} = {json.dumps(value, ensure_ascii=False)}\n" for key, value in item.items()
            )
        declaration_text = "".join(
            f"{key} = {json.dumps(value, ensure_ascii=False)}\n"
            for key, value in {
                "schema_version": "glr.season.v1",
                "season_id": season,
                "ruleset_id": ruleset,
                "environment_id": project.environment_id,
                "protocol_version": project.protocol_version,
                "status": "pending",
            }.items()
        )
        season_path(project.root, relative)
        temporary: Path | None = None
        declaration_stream = declaration.open("xb")
        try:
            with declaration_stream as stream:
                stream.write(declaration_text.encode("utf-8"))
            with tempfile.NamedTemporaryFile(dir=catalog_path.parent, delete=False) as stream:
                temporary = Path(stream.name)
                stream.write(catalog_text.encode("utf-8"))
            os.replace(temporary, catalog_path)
        except BaseException:
            declaration.unlink()
            raise
        finally:
            if temporary is not None and temporary.exists():
                temporary.unlink()
    finally:
        lock_stream.close()
        lock.unlink()
    result = select_season(project, season, ruleset)
    assert result is not None
    return result

File: crates/glr-cli/src/season.rs
use std::collections::{BTreeMap, HashSet};
use std::fs::{self, OpenOptions};
use std::io::{Read, Write};
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize, de::DeserializeOwned};
use serde_json::Value;
use sha2::{Digest, Sha256};

use crate::error::{Error, Result};
use crate::project::Project;

const LIMIT: u64 = 1024 * 1024;
pub const ENVIRONMENT_KEYS: [&str; 5] = [
    "GLR_SEASON_ID",
    "GLR_RULESET_ID",
    "GLR_SEASON_CONFIG_SHA256",
    "GLR_SEASON_CONTEXT_SHA256",
    "GLR_SEASON_CONTEXT",
];

fn invalid(message: &str) -> Error {
    Error::Invalid(message.into())
}

fn identifier(value: &str) -> Result<()> {
    if value.len() > 64
        || !value.starts_with(|c: char| c.is_ascii_lowercase())
        || !value
            .chars()
            .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '_' || c == '-')
    {
        return Err(invalid(
            "season identifiers must be portable lowercase identifiers (1..64)",
        ));
    }
    Ok(())
}

pub fn config_path(root: &Path, relative: &str) -> Result<PathBuf> {
    if relative.len() > 512
        || relative.contains(['\\', ':'])
        || relative.chars().any(char::is_control)
        || relative
            .split('/')
            .any(|part| part.is_empty() || part == "." || part == "..")
    {
        return Err(invalid(
            "season config must be a portable root-relative path",
        ));
    }
    let mut path = root.to_path_buf();
    for part in relative.split('/') {
        path.push(part);
        if let Ok(metadata) = fs::symlink_metadata(&path) {
            #[cfg(windows)]
            let reparse = {
                use std::os::windows::fs::MetadataExt;
                metadata.file_attributes() & 0x400 != 0
            };
            #[cfg(not(windows))]
            let reparse = false;
            if metadata.is_symlink() || reparse {
                return Err(invalid(
                    "season config cannot contain links or reparse points",
                ));
            }
        }
    }
    Ok(path)
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct Reference {
    pub config: String,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
struct Entry {
    season_id: String,
    ruleset_id: String,
    config: String,
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct Catalog {
    schema_version: String,
    entries: Vec<Entry>,
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct Declaration {
    schema_version: String,
    season_id: String,
    ruleset_id: String,
    environment_id: String,
    protocol_version: String,
    status: String,
    #[serde(default)]
    extensions: BTreeMap<String, Reference>,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct Input {
    pub path: String,
    pub sha256: String,
    pub size_bytes: usize,
}

fn read(root: &Path, relative: &str) -> Result<(Input, Vec<u8>)> {
    let path = config_path(root, relative)?;
    if !path.is_file() {
        return Err(invalid("season config must be an existing regular file"));
    }
    let mut bytes = Vec::new();
    fs::File::open(path)?
        .take(LIMIT + 1)
        .read_to_end(&mut bytes)?;
    if bytes.len() as u64 > LIMIT {
        return Err(invalid("season input exceeds the 1 MiB limit"));
    }
    Ok((
        Input {
            path: relative.into(),
            sha256: format!("{:x}", Sha256::digest(&bytes)),
            size_bytes: bytes.len(),
        },
        bytes,
    ))
}

fn read_toml<T: DeserializeOwned>(root: &Path, relative: &str) -> Result<(Input, T)> {
    if !relative.ends_with(".toml") {
        return Err(invalid("season declarations and catalogs must be TOML"));
    }
    let (input, bytes) = read(root, relative)?;
    let text = std::str::from_utf8(&bytes).map_err(|_| invalid("season TOML must be UTF-8"))?;
    let value = toml::from_str(text)
        .map_err(|_| invalid("season TOML has invalid, missing or unknown fields"))?;
    Ok((input, value))
}

fn catalog(project: &Project) -> Result<(Input, Catalog)> {
    let relative = project
        .seasons
        .as_deref()
        .ok_or_else(|| invalid("project has no [seasons] config reference"))?;
    let (input, catalog): (_, Catalog) = read_toml(&project.root, relative)?;
    if catalog.schema_version != "glr.seasons.v1" || catalog.entries.len() > 256 {
        return Err(invalid(
            "unsupported season catalog schema or more than 256 entries",
        ));
    }
    let mut pairs = HashSet::new();
    for entry in &catalog.entries {
        identifier(&entry.season_id)?;
        identifier(&entry.ruleset_id)?;
        config_path(&project.root, &entry.config)?;
        if !pairs.insert((&entry.season_id, &entry.ruleset_id)) {
            return Err(invalid("duplicate season/ruleset selection"));
        }
    }
    Ok((input, catalog))
}

#[derive(Debug, Clone, Serialize)]
pub struct Context {
    schema_version: &'static str,
    pub season_id: String,
    pub ruleset_id: String,
    pub environment_id: String,
    pub protocol_version: String,
    pub status: String,
    pub project: Input,
    pub catalog: Input,
    pub declaration: Input,
    pub extensions: BTreeMap<String, Input>,
}

impl Context {
    pub fn value(&self) -> Result<Value> {
        // Value objects serialize with sorted keys, matching Python's canonical encoding.
        let mut value = serde_json::to_value(self)?;
        let hash = format!("{:x}", Sha256::digest(serde_json::to_vec(&value)?));
        value
            .as_object_mut()
            .expect("context is an object")
            .insert("context_sha256".into(), hash.into());
        Ok(value)
    }

    pub fn json(&self) -> Result<String> {
        let text = serde_json::to_string(&self.value()?)?;
        if text.len() > 24 * 1024 {
            return Err(invalid(
                "season context exceeds the 24 KiB role-environment limit",
            ));
        }
        Ok(text)
    }

    pub fn verify(&self, root: &Path) -> Result<()> {
        for reference in [&self.project, &self.catalog, &self.declaration]
            .into_iter()
            .chain(self.extensions.values())
        {
            if read(root, &reference.path)?.0 != *reference {
                return Err(invalid("frozen season input changed before role execution"));
            }
        }
        Ok(())
    }

    pub fn require_ready(&self) -> Result<()> {
        if self.status != "ready" {
            return Err(invalid(
                "selected season is pending; train/goal/play require ready configuration",
            ));
        }
        Ok(())
    }
}

pub fn select(
    project: &Project,
    season: Option<&str>,
    ruleset: Option<&str>,
) -> Result<Option<Context>> {
    let (season, ruleset) = match (season, ruleset) {
        (None, None) => return Ok(None),
        (Some(season), Some(ruleset)) => (season, ruleset),
        _ => return Err(invalid("--season and --ruleset must be supplied together")),
    };
    identifier(season)?;
    identifier(ruleset)?;
    let (catalog_input, catalog) = catalog(project)?;
    let entry = catalog
        .entries
        .iter()
        .find(|entry| entry.season_id == season && entry.ruleset_id == ruleset)
        .ok_or_else(|| invalid("unknown season/ruleset selection"))?;
    let (declaration_input, declaration): (_, Declaration) =
        read_toml(&project.root, &entry.config)?;
    if declaration.schema_version != "glr.season.v1"
        || declaration.season_id != season
        || declaration.ruleset_id != ruleset
        || declaration.environment_id != project.environment_id
        || declaration.protocol_version != project.protocol_version
    {
        return Err(invalid(
            "season declaration identity does not match project and selection",
        ));
    }
    if !matches!(declaration.status.as_str(), "pending" | "ready") {
        return Err(invalid("season status must be pending or ready"));
    }
    if declaration.extensions.len() > 32 {
        return Err(invalid("season extensions exceed 32 entries"));
    }
    let mut extensions = BTreeMap::new();
    for (namespace, reference) in declaration.extensions {
        identifier(&namespace)?;
        extensions.insert(namespace, read(&project.root, &reference.config)?.0);
    }
    let manifest_name = project
        .manifest_path
        .file_name()
        .and_then(|name| name.to_str())
        .ok_or_else(|| invalid("manifest filename must be UTF-8"))?;
    let project_input = read(&project.root, manifest_name)?.0;
    if project_input.sha256 != project.manifest_sha256 {
        return Err(invalid("project manifest changed while selecting season"));
    }
    let context = Context {
        schema_version: "glr.season-context.v1",
        season_id: season.into(),
        ruleset_id: ruleset.into(),
        environment_id: project.environment_id.clone(),
        protocol_version: project.protocol_version.clone(),
        status: declaration.status,
        project: project_input,
        catalog: catalog_input,
        declaration: declaration_input,
        extensions,
    };
    context.json()?;
    Ok(Some(context))
}

pub fn list(project: &Project) -> Result<Vec<Value>> {
    if project.seasons.is_none() {
        return Ok(Vec::new());
    }
    catalog(project)?
        .1
        .entries
        .iter()
        .map(|entry| {
            select(project, Some(&entry.season_id), Some(&entry.ruleset_id))?
                .expect("selected entry")
                .value()
        })
        .collect()
}

pub fn require_selection(project: &Project, ready: bool) -> Result<()> {
    if project.seasons.is_some() && project.season_context.is_none() {
        return Err(invalid(
            "project requires explicit --season and --ruleset before role execution",
        ));
    }
    if let Some(context) = &project.season_context {
        context.verify(&project.root)?;
        if ready {
            context.require_ready()?;
        }
    }
    Ok(())
}

pub fn initialize(project: &Project, season: &str, ruleset: &str) -> Result<Value> {
    identifier(season)?;
    identifier(ruleset)?;
    let relative = project
        .seasons
        .as_deref()
        .ok_or_else(|| invalid("first declare [seasons] config in the project"))?;
    if !relative.ends_with(".toml") {
        return Err(invalid("season catalog must be TOML"));
    }
    let path = config_path(&project.root, relative)?;
    let original = if path.exists() {
        Some(read(&project.root, relative)?.1)
    } else {
        None
    };
    let mut entries = if original.is_some() {
        catalog(project)?.1.entries
    } else {
        Vec::new()
    };
    if entries
        .iter()
        .any(|entry| entry.season_id == season && entry.ruleset_id == ruleset)
    {
        return Err(invalid("season/ruleset already registered"));
    }
    if entries.len() >= 256 {
        return Err(invalid("season catalog exceeds 256 entries"));
    }
    let declaration_relative = format!("config/seasons/{season}/{ruleset}.toml");
    if declaration_relative == relative
        || (cfg!(windows) && declaration_relative.eq_ignore_ascii_case(relative))
    {
        return Err(invalid("catalog and declaration paths must differ"));
    }
    let declaration_path = config_path(&project.root, &declaration_relative)?;
    fs::create_dir_all(declaration_path.parent().expect("declaration parent"))?;
    fs::create_dir_all(path.parent().expect("catalog parent"))?;
    let lock_path = path.with_extension("toml.lock");
    let lock = OpenOptions::new()
        .write(true)
        .create_new(true)
        .open(&lock_path)?;
    let result = (|| {
        config_path(&project.root, relative)?;
        let current = if path.exists() {
            Some(read(&project.root, relative)?.1)
        } else {
            None
        };
        if current != original {
            return Err(invalid("season catalog changed during initialization"));
        }
        entries.push(Entry {
            season_id: season.into(),
            ruleset_id: ruleset.into(),
            config: declaration_relative.clone(),
        });
        let mut catalog_text = "schema_version = \"glr.seasons.v1\"\n".to_owned();
        for entry in &entries {
            catalog_text.push_str(&format!(
                "\n[[entries]]\nseason_id = {}\nruleset_id = {}\nconfig = {}\n",
                serde_json::to_string(&entry.season_id)?,
                serde_json::to_string(&entry.ruleset_id)?,
                serde_json::to_string(&entry.config)?
            ));
        }
        let text = format!(
            "schema_version = \"glr.season.v1\"\nseason_id = {}\nruleset_id = {}\nenvironment_id = {}\nprotocol_version = {}\nstatus = \"pending\"\n",
            serde_json::to_string(season)?,
            serde_json::to_string(ruleset)?,
            serde_json::to_string(&project.environment_id)?,
            serde_json::to_string(&project.protocol_version)?
        );
        config_path(&project.root, &declaration_relative)?;
        let mut declaration_file = OpenOptions::new()
            .create_new(true)
            .write(true)
            .open(&declaration_path)?;
        let write_result = (|| {
            declaration_file.write_all(text.as_bytes())?;
            drop(declaration_file);
            let mut temporary =
                tempfile::NamedTempFile::new_in(path.parent().expect("catalog parent"))?;
            temporary.write_all(catalog_text.as_bytes())?;
            temporary
                .persist(&path)
                .map_err(|error| Error::Io(error.error))?;
            Ok(())
        })();
        if write_result.is_err() {
            let _ = fs::remove_file(&declaration_path);
        }
        write_result
    })();
    drop(lock);
    let _ = fs::remove_file(lock_path);
    result?;
    select(project, Some(season), Some(ruleset))?
        .expect("created selection")
        .value()
}
